### PR TITLE
Stats thread

### PR DESCRIPTION
**Pull request by nirvn**
## Description
ATM, the statistical summary panel gathers values of given fields/expressions in the main thread, leading to a UI freeze for as long as those values are being gathered.

For large datasets, it can be pretty long (for large OGR datasets  + filter, it can be five minutes, or more).

This PR moves the value gathering off the main thread, and offers a way for the user to cancel the operation.

@nyalldawson , I had to kill QGIS a couple of times earlier today while dealing with large datasets, this would at least prevent that. It'd also prevent the whole system from being non-responsive while computing under gnome shell (grrrrrr).

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by 3nids** [comment]
maybe a silly question, but isn't this a candidate for task manager?

File: src/core/qgsvectorlayerutils.h [hunk 1]
```diff
@@ -70,6 +70,36 @@ class CORE_EXPORT QgsVectorLayerUtils
...
+
+    /**
+     * Fetches all values from a specified field name or expression.
+     * \param layer vector layer to retreive values from
+     * \param fieldOrExpression field name or an expression string
+     * \param ok will be set to false if field or expression is invalid, otherwise true
+     * \param selectedOnly set to true to get values from selected features only
+     * \param feedback optional feedback object to allow cancelation
+     * \returns list of fetched values
+     * \since QGIS 2.9
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerutils.h]: 3.0

File: src/core/qgsvectorlayerutils.h [hunk 2]
```diff
@@ -70,6 +70,36 @@ class CORE_EXPORT QgsVectorLayerUtils
         void setDuplicatedFeatures( QgsVectorLayer *layer, const QgsFeatureIds &ids );
     };
 
+    static QgsFeatureIterator getValuesIterator( const QgsVectorLayer *layer, const QString &fieldOrExpression, bool &ok, bool selectedOnly );
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerutils.h]: Missing docs

**Review by nyalldawson — COMMENTED** [review]
Looking great! Don't forget the API break docs too

**Comment by nirvn** [comment]
@nyalldawson , comments addressed.

@3nids , now using a QgsTask object.

**Comment by nirvn** [comment]
Benefit of this fix, in action:
giphy

**Comment by 3nids** [comment]
for the record, when and when not using QgsTask?

**Comment by nirvn** [comment]
@nyalldawson , awaiting your green light.

File: src/app/qgsstatisticalsummarydockwidget.cpp [hunk 3]
```diff
@@ -131,35 +142,74 @@ void QgsStatisticalSummaryDockWidget::refreshStatistics()
...
+
+    QgsApplication::taskManager()->addTask( mGatherer );
   }
 }
 
-void QgsStatisticalSummaryDockWidget::updateNumericStatistics( bool selectedOnly )
+void QgsStatisticalSummaryDockWidget::gathererFinished()
 {
-  QString sourceFieldExp = mFieldExpressionWidget->currentField();
+  mGatherer->deleteLater();
```
Comment (nyalldawson) [inline src/app/qgsstatisticalsummarydockwidget.cpp]: No need for this - task manager takes ownership, so this task will be automatically deleted.

**Pull request merged**